### PR TITLE
[Merged by Bors] - feat: define logarithmic derivatives

### PR DESCRIPTION
**Pull request by CBirkbeck**
We define the logarithmic derivative of a function as `f'/f` and then prove some basic facts about this. Specifially some things that will be needed for the Mittag-Leffler Expansion for Cotangent Function.

---

- [x] depends on: #13281 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    (hf : ∀ᶠ n : ι in atTop, DifferentiableOn ℂ (f n) s) (hg : g x ≠ 0) :
+    Tendsto (fun n : ι => logDeriv (f n) x) atTop (𝓝 ((logDeriv g) x)) := by
+  simp_rw [logDeriv]
+  apply Tendsto.div ((hF.deriv hf hs).tendsto_at x.2) (hF.tendsto_at x.2) hg
+
+section examples
+
+theorem logDeriv_sin : logDeriv (Complex.sin) = Complex.cot := by
+  rw [logDeriv, Complex.deriv_sin]
+  rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:110]: I suggest
```suggestion
  ext
  rw [logDeriv, Complex.deriv_sin, Pi.div_apply, Complex.cot]
```

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+theorem logDeriv_cos : logDeriv (Complex.cos) = -Complex.tan := by
+  simp_rw [logDeriv, Complex.deriv_cos']
+  ext x
+  simp only [Pi.div_apply]
+  exact neg_div x.cos x.sin
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:116]:
```suggestion
  ext
  rw [logDeriv, Complex.deriv_cos', Pi.div_apply, Pi.neg_apply, Complex.tan, neg_div]
```

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+/-- The derivative of `log ∘ f` is the logarithmic derivative provided `f` is differentiable and
+we are on the slitPlane. -/
+lemma deriv_clog_comp_eq_logDeriv (f : ℂ → ℂ) (x f' : ℂ) (h₁ : HasDerivAt f f' x)
+    (h₂ : f x ∈ Complex.slitPlane) : deriv (Complex.log ∘ f) x = logDeriv f x := by
+  have A := (HasDerivAt.clog h₁ h₂).deriv
+  have B := h₁.deriv
+  rw [← B] at A
+  simp only [logDeriv, Pi.div_apply, ← A]
+  rfl
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:130]:
```suggestion
  simp only [logDeriv, Pi.div_apply, ← A, Function.comp_def]
```

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 4]
```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import Mathlib.Analysis.Complex.LocallyUniformLimit
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Deriv
+import Mathlib.Analysis.SpecialFunctions.Complex.LogDeriv
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:8]: These two imports are used only for the examples; I might be inclined to put those into a separate file, probably also under `Analysis.SpecialFunctions`. This opinion is weakly held for now.

Comment (CBirkbeck) [reply Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:8]: They don't seem important enough for their own file, but I could put them at the start of the file proving the cotangent identity once its in?

Comment (urkud) [reply Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:8]: We should have a file about `logDeriv` in `Analysis/Calculus`, then we can import this file in our `SpecialFunctions` files (old ones or new ones depending on imports) and compute log derivatives of these functions there.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+theorem logDeriv_id (x : 𝕜) : logDeriv id x = 1 / x := by
+  rw [logDeriv]
+  simp only [deriv_id', Pi.div_apply, id_eq, one_div]
+
+theorem logDeriv_const (a : 𝕜') : logDeriv (fun _ : 𝕜 ↦ a) = 0 := by
+  rw [logDeriv]
+  ext1 x
+  simp only [deriv_const', Pi.div_apply, zero_div, Pi.zero_apply]
+
+theorem LogDeriv_mul {f g : 𝕜 → 𝕜'} (x : 𝕜) (hfg : f x * g x ≠ 0) (hdf : DifferentiableAt 𝕜 f x)
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:48]: Please use `logDeriv_`, not `LogDeriv_` in lemma names.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+/-- The logarithmic derivative of a function defined as `deriv f /f`. Note that it will be zero
+at `x` if `f` is not DifferentiableAt `x`. -/
+def logDeriv (f : 𝕜 → 𝕜') :=
+  deriv f / f
+
+lemma logDeriv_eq_zero_of_not_differentiableAt (f : 𝕜 → 𝕜') (x : 𝕜) (h : ¬DifferentiableAt 𝕜 f x) :
+    logDeriv f x = 0 := by
+  simp only [logDeriv, Pi.div_apply, deriv_zero_of_not_differentiableAt h, zero_div]
+
+theorem logDeriv_id (x : 𝕜) : logDeriv id x = 1 / x := by
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:39]: This should be a `@[simp]` lemma.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  rw [logDeriv]
+  simp only [deriv_id', Pi.div_apply, id_eq, one_div]
+
+theorem logDeriv_const (a : 𝕜') : logDeriv (fun _ : 𝕜 ↦ a) = 0 := by
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:43]: And this too.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  ext1 x
+  simp only [deriv_const', Pi.div_apply, zero_div, Pi.zero_apply]
+
+theorem LogDeriv_mul {f g : 𝕜 → 𝕜'} (x : 𝕜) (hfg : f x * g x ≠ 0) (hdf : DifferentiableAt 𝕜 f x)
+    (hdg : DifferentiableAt 𝕜 g x) :
+    logDeriv (fun z => f z * g z) x = logDeriv f x + logDeriv g x := by
+  simp only [logDeriv, Pi.div_apply, deriv_mul hdf hdg]
+  field_simp [(mul_ne_zero_iff.1 hfg).1, (mul_ne_zero_iff.1 hfg).2, mul_comm]
+
+theorem LogDeriv_mul_const {f : 𝕜 → 𝕜'} (x : 𝕜) (a : 𝕜') (hf : f x * a ≠ 0)
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:54]: It looks like you only need `a ≠ 0` for this lemma. If `f x = 0` or `f` is not differentiable at `x`, then both sides are zero.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    (hdf : DifferentiableAt 𝕜 f x) : logDeriv (fun z => f z * a) x = logDeriv f x := by
+  rw [LogDeriv_mul x hf hdf]
+  simp only [logDeriv_const, Pi.zero_apply, add_zero]
+  fun_prop
+
+theorem LogDeriv_const_mul {f : 𝕜 → 𝕜'} (x : 𝕜) (a : 𝕜') (hf : a * f x ≠ 0)
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:60]: Same here.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  ext
+  rw [logDeriv, Complex.deriv_cos', Pi.div_apply, Pi.neg_apply, Complex.tan, neg_div]
+
+end examples
+
+section clog
+
+/-- The derivative of `log ∘ f` is the logarithmic derivative provided `f` is differentiable and
+we are on the slitPlane. -/
+lemma deriv_clog_comp_eq_logDeriv (f : ℂ → ℂ) (x f' : ℂ) (h₁ : HasDerivAt f f' x)
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:123]: Please rename to `Complex.deriv_log_comp_eq_logDeriv` and add a `Real` version. Also please assume `DifferentiableAt`, not `HasDerivAt` and make arguments `f` and `x` implicit.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+      exact hf.2
+    · apply hd
+      simp only [Finset.cons_eq_insert, Finset.mem_insert, eq_self_iff_true, true_or_iff]
+    · rw [Finset.prod_fn]
+      apply DifferentiableAt.finset_prod
+      intro r hr
+      apply hd
+      simp only [Finset.cons_eq_insert, Finset.mem_insert, hr, or_true]
+
+theorem logDeriv_comp (f g : 𝕜 → 𝕜) (x : 𝕜) (hf : DifferentiableAt 𝕜 f (g x))
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:93]: Please make `f`, `g`, and `x` implicit.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  fun_prop
+
+theorem LogDeriv_const_mul {f : 𝕜 → 𝕜'} (x : 𝕜) (a : 𝕜') (hf : a * f x ≠ 0)
+    (hdf : DifferentiableAt 𝕜 f x) : logDeriv (fun z => a * f z) x = logDeriv f x := by
+  rw [LogDeriv_mul x hf _ hdf]
+  simp only [logDeriv_const, Pi.zero_apply, zero_add]
+  fun_prop
+
+/-- The logarithmic derivative of a finite product is the sum of the logarithmic derivatives. -/
+theorem logDeriv_prod {α : Type*} (s : Finset α) (f : α → 𝕜 → 𝕜') (t : 𝕜) (hf : ∀ x ∈ s, f x t ≠ 0)
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:67]: Please add `logDeriv_pow` and `logDeriv_zpow`.

File: Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean (hunk 4 above)
Comment (CBirkbeck) [reply Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:8]: Sorry, I don't quite understand. Do you mean make a file about `logDeriv` in `Analysis/Calculus`?

Comment (urkud) [reply Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean:8]: Yes, please move basic facts that are true in any nontrivially normed field to `Analysis/Calculus/LogDeriv`.

Comment (CBirkbeck) [reply Mathlib/Analysis/NormedSpace/LogarithmicDerivatives.lean]: Ok so I did the following: I made a new file `Analysis/Calculus/LogDeriv` and put everything that was true for nontrivially normed fields (which was most of the stuff in the file). I then made a file in `SpecialFunctions/Trigonometric` with the log derivs of the trig functions. I left the rest in the `LogarithmicDerivatives` file which I've renamed to `LogDeriv` for consistency. 

Hopefully this is what you meant? But let me know if not.

We also already have a file `Analysis/SpecialFunctions/Complex/LogDeriv` which could maybe contain the link between LodDerivs and the derivative of composing with log, but I've not modified it for the moment as I'm not sure if we want to add more imports to this file.

**Comment by urkud** [comment]
I've checked out this branch locally and working on it.

File: Mathlib/Analysis/NormedSpace/LogDeriv.lean [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+section log
+
+/-- The derivative of `log ∘ f` is the logarithmic derivative provided `f` is differentiable and
+we are on the slitPlane. -/
+lemma Complex.deriv_log_comp_eq_logDeriv {f : ℂ → ℂ} {x : ℂ} (h₁ : DifferentiableAt ℂ f x)
+    (h₂ : f x ∈ Complex.slitPlane) : deriv (Complex.log ∘ f) x = logDeriv f x := by
+  have A := (HasDerivAt.clog h₁.hasDerivAt h₂).deriv
+  rw [← h₁.hasDerivAt.deriv] at A
+  simp only [logDeriv, Pi.div_apply, ← A, Function.comp_def]
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogDeriv.lean:34]: Please move this and the next lemma to the files, where `HasDerivAt.clog` and `deriv.log` come from.

File: Mathlib/Analysis/NormedSpace/LogDeriv.lean [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  rfl
+
+end log
+
+/-- The logarithmic derivative of a sequence of functions converging locally uniformly to a
+function is the logarithmic derivative of the limit function-/
+theorem logDeriv_tendsto {ι : Type*} {p : Filter ι} (f : ι → ℂ → ℂ) (g : ℂ → ℂ)
+    {s : Set ℂ} (hs : IsOpen s) (x : s) (hF : TendstoLocallyUniformlyOn f g p s)
+    (hf : ∀ᶠ n : ι in p, DifferentiableOn ℂ (f n) s) (hg : g x ≠ 0) :
+    Tendsto (fun n : ι => logDeriv (f n) x) p (𝓝 ((logDeriv g) x)) := by
```
Comment (urkud) [inline Mathlib/Analysis/NormedSpace/LogDeriv.lean:50]: This theorem has much heavier dependencies than the previous two. Please move it to the file with `TendstoLocallyUniformlyOn.deriv`.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/LogDeriv.lean [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+import Mathlib.Analysis.Calculus.LogDeriv
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Deriv
+
+/-!
+# Logarithmic Derivatives of Trigonometric Functions
+
+We compute the logarithmic derivatives of the sine and cosine functions.
+-/
+
+theorem logDeriv_sin : logDeriv (Complex.sin) = Complex.cot := by
```
Comment (urkud) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/LogDeriv.lean:16]: I think that it's OK to `import` the new `Calculus.LogDeriv` file in existing `SpecialFunctions.*` files and add `Complex.logDeriv_sin`, `Complex.logDeriv_cos`, `Complex.logDeriv_exp`, and similarly for `Real.*`.

**Comment by urkud** [comment]
Otherwise LGTM. Thanks!

**Comment by CBirkbeck** [comment]
Thanks for the gold/cleanup! I've made the suggested changes and removed the now obsolete files.

**Comment by urkud** [comment]
Next time please remove the `awaiting-author` label once the PR is ready for another review.

File: Mathlib/Analysis/SpecialFunctions/Complex/LogDeriv.lean [hunk 16]
```diff
@@ -4,9 +4,11 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes, Abhimanyu Pallavi Sudhir, Jean Lo, Calle Sönne, Benjamin Davidson
 -/
 import Mathlib.Analysis.Calculus.InverseFunctionTheorem.Deriv
+import Mathlib.Analysis.Calculus.LogDeriv
 import Mathlib.Analysis.SpecialFunctions.Complex.Log
 import Mathlib.Analysis.SpecialFunctions.ExpDeriv
 
+
```
Comment (urkud) [inline Mathlib/Analysis/SpecialFunctions/Complex/LogDeriv.lean:11]:
```suggestion
```

**Comment by urkud** [comment]
Thanks! Once CI is happy, please merge this PR using `bors r+`
bors d+

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**